Abstract POSIX filesystem functions

zoraaver opened the pull request:
To allow non-POSIX platforms such as Windows to support WASI libc
filesystem functionality, create a set of wrapper functions which
provide a platform-agnostic interface to interact with the host
filesystem. For now, the Windows implementation is stubbed but this will
be implemented properly in a future PR. There are no functional changes
in this change, just a reorganization of code to move any direct POSIX
references out of posix.c in the libc implementation into posix_file.c
under the shared POSIX sources.

See https://github.com/bytecodealliance/wasm-micro-runtime/issues/2495 for a
more detailed overview of the plan to port the WASI libc filesystem to Windows.

zoraaver commented on core/iwasm/include/wasm_export.h:
```diff
@@ -431,26 +431,31 @@ wasm_runtime_get_module_hash(wasm_module_t module);
                            const char *env[], uint32_t env_count,
                            char *argv[], int argc,
-                           int stdinfd, int stdoutfd, int stderrfd);
+                           uint64_t stdinfd, uint64_t stdoutfd, uint64_t stderrfd);
```
This is a bit awkward because we allow the embedder to pass in custom handles for the STDIO streams. But the type of a handle is necessarily platform-dependent. i.e. on POSIX, it's `int` and on Windows it's `HANDLE`. To get around this, I've allowed the largest possible type and then we cast back to the proper type for the platform in the implementation. The embedder would also have to cast to `uint64_t` which might be annoying.

zoraaver commented on core/shared/platform/windows/platform_internal.h:
```diff
@@ -137,6 +137,12 @@ bh_atomic_thread_fence(int mem_order);
+
+typedef HANDLE os_handle;
+typedef void os_dir_stream;
+typedef HANDLE os_raw_handle;
```
`os_raw_handle` is the same type `os_handle` but that won't be true when this is actually implemented for Windows because we'll need to store some other info in `os_handle` which we can't get easily from the native `HANDLE`.

zoraaver commented on core/shared/platform/android/platform_internal.h:
```diff
@@ -140,15 +140,15 @@ seekdir(DIR *__dir, long __location);
 #endif
 
-#if __ANDROID_API__ < 24
-
```
This change is unrelated to the PR - just to fix an android build failure which has already been fixed on main. https://github.com/bytecodealliance/wasm-micro-runtime/pull/2594

zoraaver commented on .github/workflows/coding_guidelines.yml:
```diff
@@ -16,7 +16,7 @@ concurrency:
 jobs:
   compliance_job:
-    runs-on: ubuntu-latest
+    runs-on: ubuntu-20.04
```
https://github.com/bytecodealliance/wasm-micro-runtime/pull/2599#issuecomment-1739125944

loganek commented on core/iwasm/common/wasm_runtime_common.c:
```diff
@@ -2745,7 +2745,8 @@ wasm_runtime_set_wasi_args_ex(WASMModuleCommon *module, const char *dir_list[],
                               uint32 map_dir_count, const char *env_list[],
                               uint32 env_count, char *argv[], int argc,
-                              int stdinfd, int stdoutfd, int stderrfd)
+                              int64_t stdinfd, int64_t stdoutfd,
```
are file descriptors in WASI represented by int64?

loganek commented on core/iwasm/libraries/libc-wasi/sandboxed-system-primitives/src/posix.c:
```diff
@@ -587,140 +482,143 @@ fd_table_detach(struct fd_table *ft, __wasi_fd_t fd, struct fd_object **fo)
-fd_number(const struct fd_object *fo)
+// underlying os handle associated.
+static os_handle
+fd_get_handle(const struct fd_object *fo)
```
given that we no longer do the validation, and the function doesn't provide any abstraction (since `os_handle` is public), I don't know whether there's much value in keeping it - but up to you.

zoraaver replied:
Yup, we can get rid of it 👍

zoraaver commented on core/iwasm/libraries/libc-wasi/sandboxed-system-primitives/src/posix.c:
```diff
@@ -1420,21 +1151,18 @@ wasmtime_ssp_fd_write(
     ssize_t len = 0;
     /* redirect stdout/stderr output to BH_VPRINTF function */
-    if (fd_number(fo) == 1 || fd_number(fo) == 2) {
+    if (fd_get_handle(fo) == 1 || fd_get_handle(fo) == 2) {
```
Ah you're right - this is incorrect. Thanks for spotting! I'll change to use `os_get_std..` functions

wenyongh commented:
@zoraaver, @loganek There are lots of new commits in the main branch since the last synchronization, and it also resolved some CI issues, would you mind merging the main branch into this branch? If no, I will do it soon.

loganek commented:
@wenyongh thanks for flagging that, I've opened a PR: https://github.com/bytecodealliance/wasm-micro-runtime/pull/2625

wenyongh commented:
> @wenyongh thanks for flagging that, I've opened a PR: #2625

Thanks a lot! Merging the commits from the branch of your repo to the WAMR dev/wasi-libc-windows branch may introduce new commits that are ahead of the WAMR main branch, and there will be many conflicts when we merge dev/wasi-libc-windows into main in the future (since some similar changes are applied in both WAMR main and dev/wasi-libc-windows). How about I merge WAMR main into dev/wasi-libc-windows offline by using all the files of the branch of your repo to resolve the conflicts? We usually merge main into the development branches to do the synchronization.

loganek commented:
Hi @wenyongh , I've had exact clone of the WAMR repos in my fork, so I think it should be safe to merge it as it is, but let me know if I missed anything.

wenyongh commented:
> Hi @wenyongh , I've had exact clone of the WAMR repos in my fork, so I think it should be safe to merge it as it is, but let me know if I missed anything.

Hi, yes, the commits logs of your fork are same as the WAMR main branch, but the issue is that we rebase dev/wasi-libc-windows with your branch, but not WAMR main, dev/wasi-libc-windows doesn't create the new synchronization info about the last commit of WAMR main it has ever merged. Normally we will periodically merge WAMR main into dev/wasi-libc-windows, so when next time I do it, these commit logs will be merged again, and since there are similar changes in both dev/wasi-libc-windows and WAMR main, and further changes may be applied in dev/wasi-libc-windows, conflicts may occurs again. And if we don't merge WAMR main into dev/wasi-libc-windows, instead we merge your fork into dev/wasi-lib-windows each time, then conflicts will also occur in the future when we merge dev/wasi-libc-windows into WAMR main. So my preference is that we merge WAMR main into dev/wasi-libc-windows, just like we usually do.

I can also merge your PR, but as mentioned above, we had better merge WAMR main into dev/wasi-libc-windows and use exactly the same files of your repo to resolve conflicts, but I am not sure whether there will be duplicated commit logs showed after we do it, since we merge these logs twice (one from your repo, one from WAMR main).

loganek commented:
I see, thanks. Would you give me specific instructions how to prepare the merge so I can do it? I already know how to resolve the conflicts so it shouldn't take too long.

wenyongh commented:
> I see, thanks. Would you give me specific instructions how to prepare the merge so I can do it? I already know how to resolve the conflicts so it shouldn't take too long.

No, I think you have helped resolve the conflicts and that's enough, what I need to do is to clone the WAMR repo, checkout branch dev/wasi-libc-windows and "git merge main", and then clone your repo in other place, checkout your branch and copy all the files of your branch to WAMR repo to resolve the conflicts and then "git push". It won't take long.

If there are conflicts next time when we merge main into dev/wasi-libc-windows, it would be appreciated if you help resolve the conflicts and submit PR again.

loganek commented:
OK, thanks a lot. Let me know if there's anything else I can help with.

wenyongh commented:
> @wenyongh thanks for rebasing the base branch. I've resolved all conflicts in this PR and the build is now green. Would you mind taking a look when you get a chance? thanks!

@zoraaver sure, there are lots of changes, I need some time to look into it. Thanks.

wenyongh commented:
@yamt could you help review this PR? It has some changes related to libc wasi posix.c and the wakeup blocking operations. Thanks.

zoraaver commented on core/shared/platform/windows/platform_internal.h:
```diff
@@ -137,6 +137,10 @@ bh_atomic_thread_fence(int mem_order);
 
 #define os_atomic_thread_fence bh_atomic_thread_fence
 
+typedef HANDLE os_handle;
```
I'm not sure about renaming it to `os_file_handle`, the handle could also refer to a socket e.g. https://github.com/zoraaver/wasm-micro-runtime/blob/posix-filesystem-functions/core/iwasm/libraries/libc-wasi/sandboxed-system-primitives/src/posix.c#L2499.

wenyongh replied:
My understanding is that socket fd is also a kind of file descriptor, see the manual of socket function:
https://man7.org/linux/man-pages/man2/socket.2.html
```
socket() creates an endpoint for communication and returns a file descriptor that refers to that endpoint.
The file descriptor returned by a successful call will be the lowest-numbered file descriptor not currently open for the process.
```

wenyongh replied:
And in your code, after we get the file type with os_fstat, we check whether the fd is __WASI_FILETYPE_SOCKET_DGRAM or __WASI_FILETYPE_SOCKET_STREAM:
https://github.com/zoraaver/wasm-micro-runtime/blob/posix-filesystem-functions/core/iwasm/libraries/libc-wasi/sandboxed-system-primitives/src/posix.c#L508-L544

zoraaver replied:
Ok yup, that makes sense. Will change to `os_file_handle` and `os_raw_file_handle`.

zoraaver commented on core/shared/platform/include/platform_api_extension.h:
```diff
@@ -1111,6 +1111,451 @@ os_socket_get_broadcast(bh_socket_t socket, bool *is_enabled);
+ * @param flags a pointer in which to store the output
+ */
+__wasi_errno_t
+os_handle_get_fdflags(os_handle handle, __wasi_fdflags_t *flags);
```
Similar to the previous comment, the input handle might refer to a socket so I'm not sure.

zoraaver replied:
Changed 👍

zoraaver commented on core/iwasm/libraries/libc-wasi/sandboxed-system-primitives/src/blocking_op.c:
```diff
@@ -8,119 +8,67 @@
     if (!wasm_runtime_begin_blocking_op(exec_env)) {
-        errno = EINTR;
-        return -1;
+        return __WASI_EINTR;
```
I'm not sure I completely understand - where these blocking op (file) functions are called, we don't check `errno` after the function call since we can obtain the error directly from the return value of the function. But don't mind either way really, just don't know if it has any use setting errno here.

wenyongh replied:
I am not sure, the code was added by yamt in PR https://github.com/bytecodealliance/wasm-micro-runtime/pull/2516/files, I think it was needed by him. But now it seems good to directly return __WASI_EINTR.

zoraaver replied:
Ok I'll resolve this then if that's ok. Seems like we agree it's sufficient (and probably cleaner) to propagate the error via the return value rather than errno.

zoraaver commented on core/iwasm/libraries/libc-wasi/sandboxed-system-primitives/src/posix.c:
```diff
@@ -470,15 +366,13 @@ fd_prestats_remove_entry(struct fd_prestats *pt, __wasi_fd_t fd)
     struct refcount refcount;
     __wasi_filetype_t type;
-    int number;
+    os_handle os_handle;
```
Similar to https://github.com/bytecodealliance/wasm-micro-runtime/pull/2585#discussion_r1356511107 I guess, the handle might be a socket handle so not sure if renaming to `file_handle` would be suitable.

zoraaver commented on core/iwasm/libraries/libc-wasi/sandboxed-system-primitives/src/posix.c:
```diff
@@ -606,151 +500,139 @@ fd_table_detach(struct fd_table *ft, __wasi_fd_t fd, struct fd_object **fo)
+            *rights_inheriting = RIGHTS_REGULAR_FILE_INHERITING;
+            break;
+        case __WASI_FILETYPE_SOCKET_DGRAM:
+        case __WASI_FILETYPE_SOCKET_STREAM:
```
Yup, that's all checked (sorry the code has all moved around a lot!). 

So we get the filetype via `os_fstat` [here](https://github.com/zoraaver/wasm-micro-runtime/blob/posix-filesystem-functions/core/iwasm/libraries/libc-wasi/sandboxed-system-primitives/src/posix.c#L508) (a few lines above this code)
and in the POSIX implementation of [`os_fstat`](https://github.com/zoraaver/wasm-micro-runtime/blob/posix-filesystem-functions/core/shared/platform/common/posix/posix_file.c#L185), we call `S_ISFIFO` etc. and check the return value of `getsockopt` as before: https://github.com/zoraaver/wasm-micro-runtime/blob/posix-filesystem-functions/core/shared/platform/common/posix/posix_file.c#L95-L126.

wenyongh replied:
Got it, thanks a lot.

yamt commented:
> @yamt could you help review this PR? It has some changes related to libc wasi posix.c and the wakeup blocking operations. Thanks.

i will take a look later.

yamt commented on core/shared/platform/include/platform_api_extension.h:
```diff
@@ -1111,6 +1111,451 @@ os_socket_get_broadcast(bh_socket_t socket, bool *is_enabled);
+/**
+ * Returns an internally used WASI libc handle for the given raw handle to
+ * STDOUT. If the input handle is invalid (e.g. a file descriptor with a value
+ * of -1 on Linux), it will default to the platform-defined STDOUT.
```
this interface mixes a few separate functionalities:
* "is invalid" check
* type conversion
* get stdin

i feel it's more straightforward to have separate apis.

zoraaver replied:
Yes, you're right. Unfortunately C doesn't have default parameters for functions 😄 which is really what I wanted in this function.

I'll split it up into two separate functions:
`os_convert_raw_handle` to convert a `os_raw_handle` into an `os_handle`
and `os_get_stdout_handle` etc. but without any arguments. Then we can explicitly check when initializing WASI whether the user-provided handle is valid and use the platform default stdio handles if it is.

zoraaver commented on core/iwasm/libraries/libc-wasi/sandboxed-system-primitives/src/blocking_op.c:
```diff
@@ -8,119 +8,67 @@
     if (!wasm_runtime_begin_blocking_op(exec_env)) {
-        errno = EINTR;
-        return -1;
+        return __WASI_EINTR;
```
Resolving this due to https://github.com/bytecodealliance/wasm-micro-runtime/pull/2585#discussion_r1358043032

zoraaver commented on core/iwasm/libraries/libc-wasi/sandboxed-system-primitives/src/blocking_op.c:
```diff
@@ -8,119 +8,67 @@
     if (!wasm_runtime_begin_blocking_op(exec_env)) {
-        errno = EINTR;
-        return -1;
+        return __WASI_EINTR;
```
Resolving this due to https://github.com/bytecodealliance/wasm-micro-runtime/pull/2585#discussion_r1358043032

zoraaver commented on core/iwasm/libraries/libc-wasi/sandboxed-system-primitives/src/blocking_op.c:
```diff
@@ -8,119 +8,67 @@
     if (!wasm_runtime_begin_blocking_op(exec_env)) {
-        errno = EINTR;
-        return -1;
+        return __WASI_EINTR;
```
Resolving this due to https://github.com/bytecodealliance/wasm-micro-runtime/pull/2585#discussion_r1358043032

zoraaver commented on core/iwasm/libraries/libc-wasi/sandboxed-system-primitives/src/blocking_op.c:
```diff
@@ -8,119 +8,67 @@
     if (!wasm_runtime_begin_blocking_op(exec_env)) {
-        errno = EINTR;
-        return -1;
+        return __WASI_EINTR;
```
Resolving this due to https://github.com/bytecodealliance/wasm-micro-runtime/pull/2585#discussion_r1358043032

zoraaver commented on core/iwasm/libraries/libc-wasi/sandboxed-system-primitives/src/blocking_op.c:
```diff
@@ -207,20 +155,17 @@ blocking_op_socket_addr_resolve(wasm_exec_env_t exec_env, const char *host,
     if (!wasm_runtime_begin_blocking_op(exec_env)) {
-        errno = EINTR;
-        return -1;
+        return __WASI_EINTR;
```
Resolving this due to https://github.com/bytecodealliance/wasm-micro-runtime/pull/2585#discussion_r1358043032

zoraaver commented on core/iwasm/libraries/libc-wasi/sandboxed-system-primitives/src/posix.c:
```diff
@@ -606,151 +500,139 @@ fd_table_detach(struct fd_table *ft, __wasi_fd_t fd, struct fd_object **fo)
-static int
-fd_number(const struct fd_object *fo)
+static __wasi_errno_t
+close_non_stdio_handle(wasm_exec_env_t env, os_handle handle)
```
In what sense is it broken? At least, it seems to do what the function name says

zoraaver replied:
This PR is already quite large so leaving this for the moment if that's ok. I can open an issue if that's helpful.

zoraaver replied:
This PR is already quite large so leaving this for the moment if that's ok. I can open an issue if that's helpful.

zoraaver replied:
Okay this is fixed now - I've added a separate field `is_stdio` to `fd_object` so we know when to close the underlying file handle when releasing the fd object.

zoraaver commented on core/shared/platform/linux/platform_internal.h:
```diff
@@ -122,6 +122,10 @@ os_sigreturn();
 os_set_signal_number_for_blocking_op(int signo);
 
+typedef int os_handle;
+typedef DIR os_dir_stream;
```
In `posix.c`, we store the directory handle via a pointer and I'd prefer to keep it that way. It makes it simpler to check if it's been initialised or not. Also, I doubt any platform could define this as a fundamental type so we would pretty much always want to keep a pointer to it and not store the struct/object directly in the `fd_object` type.

zoraaver replied:
It still should be ok since we're not relying on the internal implementation of 'DIR'. We only pass it to the dirent functions in the posix file implementation.

yamt replied:
you are assuming a directory handle is a pointer to something. it might not be true for other platforms.

zoraaver replied:
I think that's a reasonable assumption but will change it to be a non-pointer type. It will require adding another function like `os_get_invalid_dir_stream`.

zoraaver commented on core/shared/platform/include/platform_api_extension.h:
```diff
@@ -1111,6 +1111,451 @@ os_socket_get_broadcast(bh_socket_t socket, bool *is_enabled);
+/****************************************************
+ *                     Section 3                    *
+ *                Filesystem support                *
+ ****************************************************/
```
Good point, will add a comment. 

Re sockets, character devices I don't think there's much special consideration. I've stuck all these functions under 'filesystem support' since that's their primary use case. The _majority_ of these functions can only be used with file handles. The exceptions are
`os_fstat`, `os_handle_get/set_flags`, `os_handle_get_access_mode`, `os_close`, `os_readv`, `os_writev` and `os_isatty`.

 In the comment, I'll mention that `os_handle` could also refer to a socket/char device and should support such operations if possible. (I've already drafted the Windows implementation for these functions and didn't run into much issue supporting sockets as well as files - just needed to add an internal 'handle_type' enum in `os_handle` to discriminate if the handle refers to a socket or a handle).

zoraaver commented on core/iwasm/include/wasm_export.h:
```diff
@@ -431,26 +431,31 @@ wasm_runtime_get_module_hash(wasm_module_t module);
-                           char *argv[], int argc,
-                           int stdinfd, int stdoutfd, int stderrfd);
+                           char *argv[], int argc, int64_t stdinfd,
+                           int64_t stdoutfd, int64_t stderrfd);
```
I guess we'll have to mention it in the release notes. @wenyongh do you have any concerns with this?

wenyongh replied:
Yes, we should mention this in the release notes, like PR https://github.com/bytecodealliance/wasm-micro-runtime/pull/2579.

yamt replied:
btw, why don't you use intptr_t?

zoraaver replied:
I guess it's possible there are platforms out there with 64-bit handles, independent of the machine architecture? It doesn't make a huge difference tbh but `intptr_t` would still require a cast on Windows when passing `HANDLE` types due to differing levels of indirection.

yamt replied:
well, i thought you were using int64_t here because it's large enough type to store a pointer, right?
it's exactly what intptr_t is for.

it seems a bit awkward to take int64_t and then cast it to intptr_t.

zoraaver replied:
ah that's my bad. The cast to `intptr_t` is not necessary - will remove it.

loganek commented on core/iwasm/common/wasm_runtime_common.h:
```diff
@@ -853,7 +853,7 @@ wasm_runtime_set_wasi_args_ex(WASMModuleCommon *module, const char *dir_list[],
                               uint32 map_dir_count, const char *env_list[],
                               uint32 env_count, char *argv[], int argc,
-                              int stdinfd, int stdoutfd, int stderrfd);
+                              int64 stdinfd, int64 stdoutfd, int64 stderrfd);
```
why don't we use `os_raw_file_handle` just like we do in `wasm_runtime_init_wasi`?

zoraaver replied:
`wasm_runtime_init_wasi` isn't part of the external API (i.e. declared in `wasm_export.h`) so we're free to use any internal types.

zoraaver commented:
@yamt @wenyongh would you mind taking another look? Many thanks

wenyongh commented on core/iwasm/libraries/libc-wasi/sandboxed-system-primitives/src/posix.c:
```diff
@@ -1744,7 +1385,11 @@ path_get(struct fd_table *curfds, struct path_access *pa, __wasi_fd_t fd,
+            error = os_close(fds[curfd--]);
+
+            if (error != __WASI_ESUCCESS) {
+                return error;
```
Should be `goto fail` to release resources.

wenyongh commented on core/shared/platform/common/posix/posix_file.c:
```diff
@@ -0,0 +1,1021 @@
+        out->st_filetype = __WASI_FILETYPE_DIRECTORY;
+    }
+    else if (S_ISFIFO(in->st_mode)) {
+
```
Had better remove unnecessary line?

wenyongh commented on core/shared/platform/common/posix/posix_file.c:
```diff
@@ -0,0 +1,1021 @@
+        out->st_filetype = __WASI_FILETYPE_SOCKET_STREAM;
+    }
+    else if (S_ISLNK(in->st_mode)) {
+
```
Same as above

wenyongh commented on core/shared/platform/common/posix/posix_file.c:
```diff
@@ -0,0 +1,1021 @@
+    for (size_t i = 0; i < iovcnt; ++i)
+        totalsize += iov[i].buf_len;
+
+    char *buf = wasm_runtime_malloc(totalsize);
```
Use BH_MALLOC instead of wasm_runtime_malloc in platform layer like other places, in the original design, platform layer is the lowest layer and should be independent from runtime, BH_MALLOC is a macro, it can be set to wasm_runtime_malloc or other API in the cmake file, and if it is not set, it is defined as os_malloc.

zoraaver replied:
Got it, will switch to BH_ macros. Thanks for explaining!

wenyongh commented on core/shared/platform/common/posix/posix_file.c:
```diff
@@ -0,0 +1,1021 @@
+    ssize_t len = pread(handle, buf, totalsize, offset);
+
+    if (len < 0) {
+        wasm_runtime_free(buf);
```
Use BH_FREE

wenyongh commented on core/shared/platform/common/posix/posix_file.c:
```diff
@@ -0,0 +1,1021 @@
+            break;
+        }
+    }
+    wasm_runtime_free(buf);
```
BH_FREE

wenyongh commented on core/shared/platform/common/posix/posix_file.c:
```diff
@@ -0,0 +1,1021 @@
+        size_t totalsize = 0;
+        for (size_t i = 0; i < iovcnt; ++i)
+            totalsize += iov[i].buf_len;
+        char *buf = wasm_runtime_malloc(totalsize);
```
Same as above, use BH_MALLOC/BH_FREE.

yamt commented on core/shared/platform/common/posix/posix_file.c:
```diff
@@ -0,0 +1,1021 @@
+
+static int stdin_fd = STDIN_FILENO;
+static int stdout_fd = STDOUT_FILENO;
+static int stderr_fd = STDERR_FILENO;
```
please don't use globals this way. it would break use cases with multiple instances.

zoraaver replied:
Good point, will refactor to avoid this. Thanks for spotting!

zoraaver replied:
I've changed this so we're no longer using global static variables.

wenyongh reviewed (commented):
LGTM

loganek commented on core/shared/platform/include/platform_api_extension.h:
```diff
@@ -1111,6 +1111,472 @@ os_socket_get_broadcast(bh_socket_t socket, bool *is_enabled);
+ * @return the invalid handle
+ */
+os_file_handle
+os_get_invalid_handle();
```
Given this function is only used to compare with other handles, I'd suggest something like:
```
bool os_is_handle_valid(os_file_handle *handle)
```
so it will work out of the box for os_file_handle that's not a primitive.

zoraaver replied:
> Given this function is only used to compare with other handles

actually there are a few places where this isn't the case so we'll need to keep both functions.

loganek commented on core/shared/platform/include/platform_api_extension.h:
```diff
@@ -1111,6 +1111,472 @@ os_socket_get_broadcast(bh_socket_t socket, bool *is_enabled);
 
+#if WASM_ENABLE_LIBC_WASI != 0 && WASM_ENABLE_UVWASI != 1
+
+#include "wasmtime_ssp.h"
```
I wonder if adding this file doesn't introduce a logical circular dependency - we add dependency to a header that is part of libc-wasi implementation, but wasi-libc implementation itself depends on the platform code. I know the code compiles, just wanted to make sure we don't break anything on the conceptual level.

@wenyongh / @yamt any thoughts?

zoraaver replied:
> but wasi-libc implementation itself depends on the platform code

the wasi-libc _implementation_ (i.e. posix.c) does depend on platform code but `wasmtime_ssp.h` itself doesn't depend on any platform code - it's a header which defines the internal wasi-libc interface + set of wasi types. It doesn't depend on any internal headers (except wasm_export.h which itself is a standalone header by necessity) so it should be safe for anything else to depend on it without introducing a circular dependency.

wenyongh replied:
Yes, that's really an issue, at least it doesn't comply with the design rule. In our original design, the platform layer is the lowest layer, it shouldn't rely on any other layers (except it can call libc API). It seems that this file requires wasi types, such as __wasi_errno_t, __wasi_filestat_t, __wasi_iovec_t and so on. My understanding is that we should also move them into the platform layer, and don't include wasmtime_ssp.h.

And there are similar issues in posix_file.h, which includes bh_common.h and wasi_errno.h.

@loganek what's your opinion?

wenyongh replied:
> > but wasi-libc implementation itself depends on the platform code
> 
> the wasi-libc _implementation_ (i.e. posix.c) does depend on platform code but `wasmtime_ssp.h` itself doesn't depend on any platform code - it's a header which defines the internal wasi-libc interface + set of wasi types. It doesn't depend on any internal headers (except wasm_export.h which itself is a standalone header by necessity) so it should be safe for anything else to depend on it without introducing a circular dependency.

Got it, how about moving wasmtime_ssp.h to this folder, remove wasm_export.h in it, and create another file head file to declare these APIs, and then include it in platform_api_extension.h to avoid change a lot in the latter?

zoraaver replied:
Yes, that sounds good to me. 

Just want to to check I'm understanding correctly - we split `wasmtime_ssp.h` into 2 parts: one header with just the WASI types and another header with the wasmtime_ssp functions since the platform layer doesn't need to know about these functions.

Something like (apologies if this isn't super clear):

`core/shared/platform/include/platform_wasi_types.h` -> standalone header containing wasi types which platform implementations as well as wasi-libc implementation can depend on. Not sure about the naming of this header.

`core/iwasm/libraries/libc-wasi/sandboxed-system-primitives/include/wasmtime_ssp.h` -> contains wasmtime_ssp API (e.g. `wasmtime_ssp_fd_renumber`, `wasmtime_ssp_fd_seek`) which only the wasi-libc implementation will depend on. (this header itself would depend on the `platform_wasi_types.h`.

The only reason I used wasi types in all the filesystem functions is because they're necessarily platform-agnostic and it feels like a lot of boilerplate to define `bh_` types which do the same thing, `bh_filestat_t`, `bh_filesize_t` etc. But that would be another option.

> And there are similar issues in posix_file.h, which includes bh_common.h and wasi_errno.h.

`bh_common.h` was necessary for `bh_memcpy_s` but I assume we can use `memcpy` instead - I don't know if there's any advantage/disadvantage. Not sure about the best place for `wasi_errno.h` yet.

wenyongh replied:
> Yes, that sounds good to me.
> 
> Just want to to check I'm understanding correctly - we split `wasmtime_ssp.h` into 2 parts: one header with just the WASI types and another header with the wasmtime_ssp functions since the platform layer doesn't need to know about these functions.
> 
> Something like (apologies if this isn't super clear):
> 
> `core/shared/platform/include/platform_wasi_types.h` -> standalone header containing wasi types which platform implementations as well as wasi-libc implementation can depend on. Not sure about the naming of this header.

How about changing to platform_wasi.h, moving the current wasi API declarations from platform_api_extension.h to it (so it includes wasi types and apis), and letting platform_api_extension.h include it?

> 
> `core/iwasm/libraries/libc-wasi/sandboxed-system-primitives/include/wasmtime_ssp.h` -> contains wasmtime_ssp API (e.g. `wasmtime_ssp_fd_renumber`, `wasmtime_ssp_fd_seek`) which only the wasi-libc implementation will depend on. This header itself would depend on `platform_wasi_types.h`.
> 

It is ok to me.

> The only reason I used wasi types in all the filesystem functions is because they're necessarily platform-agnostic and it feels like a lot of boilerplate to define `bh_` types which do the same thing, `bh_filestat_t`, `bh_filesize_t` etc. But that would be another option.

Seems better to keep unchanged (use the current __wasi prefix)?

> 
> > And there are similar issues in posix_file.h, which includes bh_common.h and wasi_errno.h.
> 
> `bh_common.h` was necessary for `bh_memcpy_s` but I assume we can use `memcpy` instead - I don't know if there's any advantage/disadvantage. Not sure about the best place for `wasi_errno.h` yet.

It should be ok to use memcpy in platform as soon as we can ensure the security. bh_memcpy_s adds more checks but  it should be not mandatory in platform layer.

loganek replied:
Yes, I agree with the approach above in general. My main problem was that we've included wasmtime_ssp.h file that was part of the wasi-libc library (even though it was just a header file).

Moving parts of it out of wasi-libc seems like a good simplification, and I agree that having multiple duplicated `bh_` types seems unnecessary.

zoraaver replied:
I've added a separate header `platform_wasi.h` as suggested. I've also moved `wasi_errno.h` to shared platform sources and renamed it to `libc_errno.h` but there might be a better place/name for that. The additional changes (moving the WASI types around) are in a separate commit to make it a bit easier to review.

let me know if that looks good and hopefully this can be merged soon enough!

zoraaver commented on core/iwasm/common/wasm_runtime_common.c:
```diff
@@ -2792,9 +2793,9 @@ wasm_runtime_set_wasi_args_ex(WASMModuleCommon *module, const char *dir_list[],
-    wasi_args->stdio[2] = stderrfd;
+    wasi_args->stdio[0] = (os_raw_handle)stdinfd;
+    wasi_args->stdio[1] = (os_raw_handle)stdoutfd;
+    wasi_args->stdio[2] = (os_raw_handle)stderrfd;
```
actually reverting this - there isn't a compilation warning and the intermediate cast to `intptr_t` is not necessarily correct since the size of a platform handle isn't always going to dictated by the pointer size on the machine.

wenyongh commented on core/iwasm/libraries/libc-wasi/sandboxed-system-primitives/include/wasmtime_ssp.h:
```diff
@@ -22,1347 +18,594 @@
 #include <stddef.h>
 #include <stdint.h>
 
+#include "platform_wasi.h"
```
Had better include "bh_patform.h", we usually include it instead of including files under core/shared/platform/include and core/shared/utils.

zoraaver replied:
changed to `bh_platform.h`, thanks

wenyongh reviewed (commented):
LGTM

loganek reviewed (approved):
LGTM.
@wenyongh if possible, could we wait with merging this PR before we merge this one first: https://github.com/bytecodealliance/wasm-micro-runtime/pull/2637 ? Many thanks.

wenyongh commented:
> LGTM. @wenyongh if possible, could we wait with merging this PR before we merge this one first: #2637 ? Many thanks.

Sure, I will merge this PR first.

@zoraaver, will you address the comment of `#include "platform_wasi.h"`? Or let's merge this PR and fix it in the next PR?

zoraaver commented:
> @zoraaver, will you address the comment of `#include "platform_wasi.h"`? Or let's merge this PR and fix it in the next PR?

addressed the comment 👍
